src/contract/link.rs

```diff
@@ -0,0 +1,233 @@
+    /// incrementally, but rather when `build` is called so that linking errors
+    /// only need to be handled in one place.
+    fn add_library(mut self, name: &str, library: Library) -> Self {
+        let name = name.to_owned();
+        self.libraries.push((name, library));
+        self
+    }
+
+    /// Links the libraries and binaries together and returns a verified
+    /// deployment.
+    ///
+    /// This method will return an error if it finds unresolved or unused
+    /// libraries during the linking process.
+    pub fn build(mut self) -> Result<Deployment, LinkerError> {
+        let mut pending_libraries = HashMap::new();
+        for (name, library) in self.libraries {
+            match library {
+                Library::Resolved(address) => self.contract_bytecode.link(&name, address)?,
+                Library::Pending(bytecode) => {
+                    // NOTE: Check that the map doesn't contain the library
+                    //   first because inserting moves `name` into the map.
+                    if pending_libraries.contains_key(&name) {
+                        return Err(LinkerError::UnusedDependency(name));
+                    }
+                    pending_libraries.insert(name, bytecode);
+                }
+            }
+        }
+
+        let mut libraries_to_deploy = Vec::new();
+        for library in self.contract_bytecode.undefined_libraries() {
+            if let Some((name, bytecode)) = pending_libraries.remove_entry(library) {
+                let bytes = match bytecode.try_to_bytes() {
+                    Some(bytes) => bytes,
+                    None => return Err(LinkerError::NestedDependencies(name)),
+                };
+                libraries_to_deploy.push((name, bytes));
+            } else {
+                return Err(LinkerError::MissingDependency(library.to_owned()));
+            }
+        }
+
+        // NOTE: At this point, the contract bytecode should be completely
+        //   linkable, as we linked all the library instance addresses and
+        //   verfied that the remaining dependencies are to be deployed. The
+        //   libraries remaning in `pending_libraries` are extra uneeded
+        //   dependencies. Report an error with the first unused dependency.
+        if let Some(unused_dependency) = pending_libraries.keys().next() {
+            return Err(LinkerError::UnusedDependency(unused_dependency.to_owned()));
+        }
+
+        Ok(Deployment {
+            libraries: libraries_to_deploy,
+            contract: (self.contract_bytecode, self.encoded_contructor_params),
+        })
+    }
+
+    /// Links the libraries and binaries together and returns a `DeployBuilder`
+    /// to setup the transaction for deploying the contract and its libraries.
+    ///
+    /// This method will return an error if it finds unresolved or unused
+    /// libraries during the linking process.
+    pub fn deploy(self) -> Result<(), LinkerError> {
+        todo!()
```

Comment:
Will this not be asynchronous?

Reply:
It will, still not sure how I want the API to be. I am struggling to find an abstraction that I am happy with.